### ISSUE TITLE
Feature request: rating for packages

### ISSUE THREAD

**Issue by flip111** [commercialhaskell/stack#3627]
Could stackage add some rating for a package? The rating could help to recommend a package. There are several metrics which could be used (i have no idea right now what combination would work out the best).

* how many other libraries are dependent on this package (higher is better)
* user voting. Could also vote on some metrics like: ease of use, documentation, performance, compatibilty, etc ..
* amount of commits
* how long has a package been in development
* amount of time with no or little commits (more time indicates "death" of a package)
* Maybe the ability to add written reviews with list of cons and pros
* How many other contributions the author of the package has made (let packages by experienced authors get a higher rating)

What would also be nice is if there are recommendations for collections of packages. For example if you have a web framework packages that the author (or someone else?) can recommend some database library or/and serialization library that plays well with this package.

I also like to see some system in place so that best practices can be established, but i think this is more long term goal (after just general rating) and perhaps also needs to be more finegrained (not on package level but on code level).

A recommendation system with some smart algorithms behind it would be awesome. I know it's a big task, but i didn't see any issue about it yet. So let's start the discussion ...

**Comment by kadoban** [comment]
I like the basic goal (trying to differentiate bad packages from good), but I think I disagree with almost every one of those criteria, unfortunately. They're either IMO not good signifiers of quality, they're horribly game-able, or they will be spammed and will have to be moderated heavily and at great expense/effort.

I question if criteria even exist which have none of those problems.

**Comment by lwm** [comment]
@flip111, have you spoken to people to hackage people? I believe they have had some rating system on the packages there and might be able to provide some insight into what is used, what is valuable, etc.

**Comment by flip111** [comment]
@kadoban not sure if you are saying if the criteria are not perfect or just bad. I know they are not perfect but i think they are pretty good. Not convinced that the bad things you mentioned are actually going to happen, perhaps i just have more faith in the haskell community.

@lwm no i haven't, which people are you talking about? I saw that hackage has a rating system but it's not very visible, near the end of the list stuck between other properties of the project. I just made the issue here as i get the impression that stack/stackage is progressing faster than hackage.

**Comment by lwm** [comment]
> no i haven't, which people are you talking about?

https://github.com/haskell/hackage-server

**Comment by flip111** [comment]
@lwm i see. Extra insight is great, but it's another project. Do you want to move efforts to there, implement something twice, have the same system for both stackage/hackage or something else? Of course you are free to open an issue on hackage-server and refer back to here if you want ^^ Maybe i will do that myself later ..

**Comment by lwm** [comment]
? 

I am recommending to you that you *look* at a Haskell project that already does package rating.

**Comment by flip111** [comment]
@lwm i saw it, example: `Rating | 2.5 (votes: 4) 	  [estimated by rule of succession]` (if this is not what you are referring to please elaborate). Whatever there is already in place .. haskell 2017 [survey results](http://taylor.fausak.me/2017/11/15/2017-state-of-haskell-survey-results/#question-37) show most people are disagreeing or neutral towards the statement `I can easily compare Haskell libraries to select the best one.`

**Comment by kadoban** [comment]
@flip111 I don't think any of those criteria are good enough to be worth using, personally. I think they are of negative value, especially without someone committing a lot of time and effort to moderating them for spam or other abuse. And a bunch of them are going to be pretty much impossible to moderate, for example it'll be really hard to tell a legitimate vote from a bogus one.

Also keep in mind that you don't just have to worry about the haskell community, anyone on the internet is going to be able to access the site. It might take general spammers a little longer to notice, but all it's going to take is a couple of them to put in a minimal amount of effort.

**Comment by flip111** [comment]
@kadoban a lot of the bullet points in the opening post don't require any moderation. As for the other bullet points ... we are not even there yet. Can deal with the problems you mentioned if they actually arise. But ok i'll bite .. it's 2017, almost 2018 how hard can it be to try to see if something is working and if not just change or remove it? Each one will be a property of the package, the logic to get the data depends on the property, which can be changed .. to .. for example .. link to an account.

**Comment by kadoban** [comment]
@flip111 You are correct, some of those don't allow/require any moderation. For example, "amount of commits". But if that becomes a signifier of quality that anybody cares about, I can guarantee you that some new packages will come along with many many commits and get to the top of the ranking. Why wouldn't they?

And then what do we do?

I would think before implementing would be the time to plan and game this out, to see if there are criteria that can be used without becoming a huge spam/malware target or at least to estimate the effort that will be required to keep the ratings somewhat legitimate. I'm not sure what 2018 has to do with it. Just implementing it and then seeing how abused it gets doesn't sound like the best idea to me personally.
